clang/tools/3c/utils/port_tools/common.py

```diff
@@ -0,0 +1,31 @@
+# Data structures that need to be imported by both generate_ccommands and
+# expand_macros.
+
+from typing import List, NamedTuple
+import functools
+import os
+
+# We are assuming it's OK to cache canonical paths for the lifetime of any
+# process that uses this code.
+realpath_cached = functools.lru_cache(maxsize=None)(os.path.realpath)
+
+class TranslationUnitInfo(NamedTuple):
+    compiler_path: str
+    # Any file paths in compiler_args (-I, etc.), input_filename, and
+    # output_filename may be relative to target_directory.
+    compiler_args: List[str]
+    target_directory: str
+    input_filename: str
+    output_filename: str
+
+    def realpath(self, path: str):
+        return realpath_cached(os.path.join(self.target_directory, path))
+
+    # Perhaps this could be cached. It's not a big cost though.
+    @property
```

Comment:
I'm not a Python programmer. What does `@property` mean?

Reply:
It means that if `tu` is a `TranslationUnitInfo` and you write `tu.input_realpath`, that becomes a call to this method, as if `@property` were not used and you wrote `tu.input_realpath()`.  So this looks like a field to clients, but the value is computed on demand.  [Documentation](https://docs.python.org/3/library/functions.html#property)